### ISSUE TITLE
feat request: "impossible comparison" diagnostic

### ISSUE THREAD

**Issue by Rudxain** [EmmyLuaLs/emmylua-analyzer-rust#879]
This code
```lua
print(0 == "a")
```
Doesn't trigger a diagnostic. TypeScript has the error `ts2367` ("types have no overlap") for this situation.

The diagnostic is more useful when comparing variables rather than literals:
```lua
---@param a int
---@param b string
function f(a, b)
  return a == b
end
```

**Comment by NyakoFox** [comment]
How would this interact with metatables?

**Comment by Rudxain** [comment]
> How would this interact with metatables?

Since ES/JS is very dynamic (perhaps even more than Lua!), I guess we could just copy whatever TS does when dealing with [proxies](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy). Even without proxies, it's possible for a `class` to define [impure](https://en.wikipedia.org/wiki/Pure_function) getters+setters that are called on property-access or assignment.

IIRC, TS simply ignores the impurity of getters/setters (for the same reason why it assumes [monkey-patching](https://en.wikipedia.org/wiki/Monkey_patch) is impossible) and prints the diagnostic anyways.

[The FAQ](https://github.com/Microsoft/TypeScript/wiki/FAQ) might have more insights on this

**Comment by xuhuanzy** [comment]
Consider diagnosing only literals.
